stackstac/rio_reader.py

```diff
@@ -399,14 +401,23 @@ def read(self, window: Window, **kwargs) -> np.ndarray:
 
             raise RuntimeError(msg) from e
 
+        if result.dtype != self.dtype:
+            result = result.astype(self.dtype, copy=False)
+
         if self.rescale:
-            scale, offset = reader.scale_offset
+            scale, offset = self.scale_offset
+
+            if np.isnan(scale) or np.isnan(offset):
+                scale, offset = reader.scale_offset
+
             if scale != 1:
+                if self.dtype in INTEGER_DTYPES:
+                    raise ValueError(f'Requested asset dtype ({self.dtype}) is not compatible with '
```

Comment:
I think the check here would need to be more sophisticated. The goal of this check is to prevent silent information loss, and alert users that they need to either not rescale, or change the output data type to something else that can fit the rescaled values.

But just checking if the output dtype is integral is too strict. Rescaling in an integer dtype _could_ be perfectly fine, if you're scaling by an integer, and the resulting values still fit in the new dtype. Rescaling from float to int could even be fine; maybe the values are all integers, but unnecessarily stored in float64 on disk. 

Ideally we'd probably temporarily set NumPy to [raise on overflow](https://numpy.org/doc/stable/reference/generated/numpy.seterr.html) and just try to do the rescale, and handle that error if it occurs, but `np.errstate` is probably not threadsafe, which this method needs to be.

Note that right now, the [docs](https://stackstac.readthedocs.io/en/latest/api/main/stackstac.stack.html#stackstac.stack.params.rescale) for `rescale` say
> Note that this could produce floating-point data when the original values are ints, so set dtype accordingly. You will NOT be warned if the cast to dtype is losing information!

which makes me feel that for this first PR, maybe it's better to not validate at all than to have overly strict validation.

Reply:
Hmm yes I see how it is overly strict. I will remove the validation.